fdbserver/KeyValueStoreRocksDB.actor.cpp

```diff
@@ -26,6 +26,42 @@
 
 namespace {
 
+class RocksDBErrorListener : public rocksdb::EventListener {
+public:
+	RocksDBErrorListener(ThreadReturnPromise<Void>* promise) : errorPromise(promise){};
+	void OnBackgroundError(rocksdb::BackgroundErrorReason reason, rocksdb::Status* bg_error) override {
+		TraceEvent(SevError, "RocksDBBGError")
+		    .detail("Reason", reason)
+		    .detail("RocksDBSeverity", bg_error->severity())
+		    .detail("Status", bg_error->ToString());
+		std::unique_lock<std::mutex> lock(mutex);
+		if (isSet)
+			return;
+		// RocksDB generates two types of background errors, IO Error and Corruption
+		// Error type and severity map could be found at
+		// https://github.com/facebook/rocksdb/blob/2e09a54c4fb82e88bcaa3e7cfa8ccbbbbf3635d5/db/error_handler.cc#L138.
+		// All background errors will be treated as storage engine failure. Send the error to storage server.
+		if (bg_error->IsIOError()) {
+			errorPromise->sendError(io_error());
+		} else if (bg_error->IsCorruption()) {
+			errorPromise->sendError(file_corrupt());
+		} else {
+			errorPromise->sendError(unknown_error());
+		}
+		isSet = true;
+	}
+	~RocksDBErrorListener() {
```

Comment:
Do we know what guarantees there are for concurrent access with the destructor? In particular, are all accesses from other threads guaranteed to be finished (with some synchronization) before the destructor gets called? To be safe, it's probably worth taking the mutex before acting on the `errorPromise`.

Reply:
Updated the code to grab the mutex first.